mem/buffers.go

```diff
@@ -0,0 +1,217 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/*
+Package mem does memory things
+*/
+package mem
+
+import (
+	"io"
+	"sync/atomic"
+)
+
+type BufferSlice []*Buffer
+
+type Buffer struct {
+	data  []byte
+	refs  *atomic.Int32
+	free  func([]byte)
+	freed bool
+}
+
+func NewBuffer(data []byte, free func([]byte)) *Buffer {
+	b := &Buffer{data: data, refs: new(atomic.Int32), free: free}
+	b.refs.Add(1)
+	return b
+}
+
+func Copy(data []byte, pool BufferPool) *Buffer {
+	buf := pool.Get(len(data))
+	copy(buf, data)
+	return NewBuffer(buf, pool.Put)
+}
+
+func (b *Buffer) ReadOnlyData() []byte {
+	if b.freed {
+		panic("Cannot read freed buffer")
+	}
+	return b.data
+}
+
+func (b *Buffer) Ref() *Buffer {
+	if b.freed {
+		panic("Cannot ref freed buffer")
+	}
+
+	b.refs.Add(1)
+	return &Buffer{
+		data: b.data,
+		refs: b.refs,
+		free: b.free,
+	}
+}
+
+func (b *Buffer) Free() {
+	if b.freed {
+		return
+	}
+
+	b.freed = true
+	refs := b.refs.Add(-1)
+	if refs == 0 && b.free != nil {
+		b.free(b.data)
+	}
+	b.data = nil
+}
+
+func (b *Buffer) Len() int {
+	// Convenience: io.Reader returns (n int, err error), and n is often checked
+	// before err is checked. To mimic this, Len() should work on nil Buffers.
+	if b == nil {
+		return 0
+	}
+	return len(b.ReadOnlyData())
+}
+
+func (b *Buffer) Split(n int) *Buffer {
+	if b.freed {
+		panic("Cannot split freed buffer")
+	}
+
+	b.refs.Add(1)
+
+	data := b.data
+	var free func([]byte)
+	if f := b.free; f != nil {
+		free = func(_ []byte) {
+			f(data)
+		}
+	}
```

Comment:
- Lets say we have a `Buffer`, `b1` and is the only ref to the underlying `[]byte`
- Now, if we do `b2 := b1.Split(n)`, the refs on both `b1` and `b2` are going to `2`
- But `b1.free` is now `nil`
- Let's say `b2` is freed first, and it decrements `refs` and since it is not yet `0`, it wont call the free function
- And when `b1` is freed, the `refs` reach `0`, but the free function is `nil`. So, the underlying `[]byte` is not freed.

Am I missing something?